modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -23,48 +37,128 @@ import eu.timepit.refined.types.numeric.{NonNegLong, PosLong}
+  def pullLatestSnapshot: F[LatestSnapshotTuple]
+  def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]]
   def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalIncrementalSnapshot]]]
   def pullGlobalSnapshot(hash: Hash): F[Option[Hashed[GlobalIncrementalSnapshot]]]
 }
 
 object GlobalL0Service {
+  val minimumMajorityPeerIds = 2
+
+  case object NoPeersWithMajorityHash extends Exception("No peers returned snapshot with hash in the majority") with NoStackTrace
+  case object NoMajoritySnapshotData extends Exception("Unable to determine latest snapshot data for majority") with NoStackTrace
+  case object NoPeerAlignedWithMajority extends Exception("No peer available that is aligned with majority") with NoStackTrace
 
   def make[
     F[_]: Async: KryoSerializer: SecurityProvider
   ](
     l0GlobalSnapshotClient: L0GlobalSnapshotClient[F],
     globalL0ClusterStorage: L0ClusterStorage[F],
     lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
-    singlePullLimit: Option[PosLong]
+    singlePullLimit: Option[PosLong],
+    maybeMajorityPeerIdSet: Option[NonEmptySet[PeerId]]
   ): GlobalL0Service[F] =
     new GlobalL0Service[F] {
 
       private val logger = Slf4jLogger.getLogger[F]
+      private val maybeMajorityPeerIds = maybeMajorityPeerIdSet.map(_.toNonEmptyList)
 
-      def pullLatestSnapshot: F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
-        globalL0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
-          l0GlobalSnapshotClient.getLatest(l0Peer).flatMap {
-            case ((snapshot, state)) =>
-              snapshot.toHashedWithSignatureCheck.flatMap(_.liftTo[F]).map((_, state))
-          }
-        }
+      def pullLatestSnapshot: F[LatestSnapshotTuple] =
+        maybeMajorityPeerIds.fold(pullLatestSnapshotFromRandomPeer)(pullLatestSnapshotWithMajorityHash)
 
       def pullGlobalSnapshot(hash: Hash): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(hash)).handleErrorWith { e =>
           logger
             .warn(e)(s"Failure pulling single snapshot with hash=$hash")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .as(none)
         }
 
       def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(ordinal)).handleErrorWith { e =>
           logger
             .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds.fold {
+          pullGlobalSnapshotsFromRandomPeer
+        } { majorityPeerIds =>
+          lastGlobalSnapshotStorage.getOrdinal.flatMap {
```

Comment:
I would extract this part into a separate function `pullGlobalSnapshotsFromMajority` similar to how you have it for `pullLatestSnapshot` right now.